### PR TITLE
[FHIR deduplication] - Encounters

### PR DESCRIPTION
**Pull request by RamilGaripov**
refs. metriport/metriport-internal#2120

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/2611

### Description
- Encounter dedup and tests
- DiagnosticReport dedup and tests

### Testing

- Local
  - [x] Unit tests

### Release Plan
- [ ] Merge this

### CODE REVIEW

File: packages/core/src/fhir-deduplication/resources/diagnostic-report.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    master.status = pickMostDescriptiveStatus(statusRanking, existing.status, target.status);
+    return master;
+  }
+
+  for (const diagReport of diagReports) {
+    const date = getDateFromResource(diagReport, "date-hm");
+
+    if (date) {
+      const key = JSON.stringify({ date });
```
Comment (jonahkaye) [inline packages/core/src/fhir-deduplication/resources/diagnostic-report.ts:91]: we gotta also thing about what is the minimum valuable resource that is clinically useful. Do we use diagnosticReports without presented forms?

File: packages/core/src/fhir-deduplication/resources/encounter.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  ): Encounter {
+    master.status = pickMostDescriptiveStatus(statusRanking, existing.status, target.status);
+    return master;
+  }
+
+  for (const encounter of encounters) {
+    const classCode = encounter.class?.code;
+    const date = getDateFromResource(encounter, "date-hm");
+    if (date && classCode) {
+      const key = JSON.stringify({ date, classCode });
```
Comment (jonahkaye) [inline packages/core/src/fhir-deduplication/resources/encounter.ts:74]: see [this](https://metriport.slack.com/archives/C04CXLSAF7V/p1724197632405539?thread_ts=1724176475.295279&cid=C04CXLSAF7V):

File: packages/core/src/fhir-deduplication/resources/diagnostic-report.ts (hunk 1 above)
Comment (jonahkaye) [reply packages/core/src/fhir-deduplication/resources/diagnostic-report.ts:91]: see [this](https://metriport.slack.com/archives/C06CQC7GQG3/p1724198701172689?thread_ts=1723806266.310919&cid=C06CQC7GQG3)

Comment (RamilGaripov) [reply packages/core/src/fhir-deduplication/resources/diagnostic-report.ts:91]: Right, I guess I did this one hastily, and forgot to do proper analysis. I'll add some comments and analysis on slack under your thread

File: packages/core/src/fhir-deduplication/__tests__/group-same-encounters.test.ts
Comment (jonahkaye) [inline packages/core/src/fhir-deduplication/__tests__/group-same-encounters.test.ts:1]: just want to make sure that the deepMerge is working and preserving the following values:
1. type
2. hospitalization.dischargeDisposition
3. reasonCode
4. condition References

**Review by jonahkaye — APPROVED** [review]
Approval pending confirmation via a test that the merge is working

File: packages/core/src/fhir-deduplication/__tests__/group-same-encounters.test.ts
Comment (RamilGaripov) [reply packages/core/src/fhir-deduplication/__tests__/group-same-encounters.test.ts:1]: addressed

**Pull request merged**